### PR TITLE
Fix COS problem

### PR DESCRIPTION
**Pull request by wangzhen11aaa**
clickhouse-backup list remote

???        28/12/2021 05:47:22   remote      broken (can't stat metadata.json)
download_daily_mv_backup8  

->

download_daily_mv_backup-8  585B       06/01/2022 12:25:26   remote      tar

---------------------

### CODE REVIEW

File: pkg/new_storage/cos.go [hunk 1]
```diff
@@ -78,30 +78,48 @@ func (c *COS) DeleteFile(key string) error {
 }
 
 func (c *COS) Walk(cosPath string, recursive bool, process func(RemoteFile) error) error {
-	prefix := path.Join(c.Config.Path, cosPath)
+	// COS needs prefix ended with "/".
+	prefix := path.Join(c.Config.Path, cosPath) + "/"
```
Comment (Slach) [inline pkg/new_storage/cos.go:82]: looks like you have wrong config
where `path` in `cos` section is empty
better will check this case or add config validation

File: pkg/new_storage/cos.go [hunk 2]
```diff
@@ -78,30 +78,48 @@ func (c *COS) DeleteFile(key string) error {
...
 	}
-	for _, v := range res.Contents {
-		modifiedTime, _ := parseTime(v.LastModified)
-		if err := process(&cosFile{
-			name:         strings.TrimPrefix(v.Key, prefix),
-			lastModified: modifiedTime,
-			size:         int64(v.Size),
-		}); err != nil {
-			return err
+	if recursive {
```
Comment (Slach) [inline pkg/new_storage/cos.go:113]: looks much better

File: pkg/new_storage/general.go [hunk 3]
```diff
@@ -135,7 +147,7 @@ func (bd *BackupDestination) BackupList() ([]Backup, error) {
 		if err != nil {
 			result = append(result, Backup{
 				metadata.BackupMetadata{
-					BackupName: strings.Trim(o.Name(), "/"),
+					BackupName: metadataFilePath,
```
Comment (Slach) [inline pkg/new_storage/general.go]: still strange backup name, I propose return `strings.Trim(o.Name(), "/")` here

File: pkg/new_storage/general.go [hunk 4]
```diff
@@ -151,7 +163,7 @@ func (bd *BackupDestination) BackupList() ([]Backup, error) {
 		if err := json.Unmarshal(b, &m); err != nil {
 			result = append(result, Backup{
 				metadata.BackupMetadata{
-					BackupName: strings.Trim(o.Name(), "/"),
+					BackupName: metadataFilePath,
```
Comment (Slach) [inline pkg/new_storage/general.go]: still strange backup name, I propose return `strings.Trim(o.Name(), "/")` here

**Review by Slach — CHANGES_REQUESTED** [review]

File: pkg/new_storage/cos.go (hunk 2 above)
Comment (wangzhen11aaa) [reply pkg/new_storage/cos.go:113]: It is not ok. Passing bool value will turn function more complex. We can optimize this later. TODO, Split function with bool variable into two functions.

File: pkg/new_storage/general.go (hunk 4 above)
Comment (wangzhen11aaa) [reply pkg/new_storage/general.go]: The core logic to get the backup is through the metadata.json file, right? The name "metadataFilePath" variable here  makes this concept clear and conceals other messy result. If you insist, I can turn this code back, That's Ok

File: pkg/new_storage/cos.go (hunk 1 above)
Comment (wangzhen11aaa) [reply pkg/new_storage/cos.go:82]: > looks like you have wrong config where `path` in `cos` section is empty better will check this case or add config validation

I use COS the first time, the path config is empty, It does not work. When check this API doc, u can find the "/" suffixed with every path in their query case. (%2F = "/", the html code).

Comment (wangzhen11aaa) [reply pkg/new_storage/cos.go:82]: > looks like you have wrong config where `path` in `cos` section is empty better will check this case or add config validation

I used COS the first time, the path config is exactly empty, It did not work either. When check this API doc, I can find the "/" suffixed with every path in their query case. (%2F = "/", the html code).

**Review by Slach — APPROVED** [review]

**Pull request merged**